operator/builtin/output/googlecloud/proto.go

```diff
@@ -40,91 +40,131 @@ func toProto(v interface{}) (*structpb.Value, error) {
 	case float64:
 		return structpb.NewNumberValue(v), nil
 	case string:
-		return toProtoString(v)
+		return convertStringToProto(v)
 	case []byte:
-		s := base64.StdEncoding.EncodeToString(v)
-		return structpb.NewStringValue(s), nil
+		return convertBytesToProto(v)
 	case map[string]interface{}:
-		v2, err := toProtoStruct(v)
-		if err != nil {
-			return nil, err
-		}
-		return structpb.NewStructValue(v2), nil
+		return convertMapToProto(v)
 	case map[string]string:
-		fields := map[string]*structpb.Value{}
-		for key, value := range v {
-			fields[key] = structpb.NewStringValue(value)
-		}
-		return structpb.NewStructValue(&structpb.Struct{
-			Fields: fields,
-		}), nil
+		return convertStringMapToProto(v)
 	case map[string]map[string]string:
-		fields := map[string]*structpb.Value{}
-		for key, value := range v {
-			proto, err := toProto(value)
-			if err != nil {
-				return nil, err
-			}
-
-			fields[key] = proto
+		return convertEmbeddedMapToProto(v)
+	case []interface{}:
+		return convertListToProto(v)
+	case []string:
+		return convertStringListToProto(v)
+	default:
+		return nil, fmt.Errorf("invalid type for proto conversion: %T", v)
+	}
+}
+
+// convertStringToProto converts a string to a proto string
+func convertStringToProto(value string) (*structpb.Value, error) {
+	if !utf8.ValidString(value) {
+		return nil, fmt.Errorf("invalid UTF-8 in string: %q", value)
+	}
+	return structpb.NewStringValue(value), nil
+}
+
+// convertBytesToProto converts a byte array to a protobuf equivalent
+func convertBytesToProto(bytes []byte) (*structpb.Value, error) {
```

Comment:
Should this return and error? It doesn't look like it can.

Reply:
No. It was probably because all the other conversion functions followed this pattern. I'll switch it to not return an error, which is probably more correct.